sgkit/stats/aggregation.py

```diff
@@ -285,6 +286,181 @@ def call_rate(ds: Dataset, dim: Dimension, call_genotype_mask: Hashable) -> Data
+)
+def _count_biallelic_genotypes(
+    genotypes: ArrayLike, _: ArrayLike, out: ArrayLike
+) -> ArrayLike:  # pragma: no cover
+    out[:] = 0
+    for i in range(len(genotypes)):
+        index = _biallelic_genotype_index(genotypes[i])
+        if index >= 0:
+            out[index] += 1
+
+
+# implementation from github.com/PlantandFoodResearch/MCHap
+# TODO: replace with math.comb when supported by numba
+@numba_jit(nogil=True)
+def _comb(n: int, k: int) -> int:
+    if k > n:
+        return 0
+    r = 1
+    for d in range(1, k + 1):
+        gcd_ = math.gcd(r, d)
+        r //= gcd_
+        r *= n
+        r //= d // gcd_
+        n -= 1
+    return r
+
+
+@numba_jit(nogil=True)
+def _comb_with_replacement(n: int, k: int) -> int:
+    n = n + k - 1
+    return _comb(n, k)
+
+
+@numba_jit(nogil=True)
+def _sorted_genotype_index(genotype: ArrayLike) -> int:
+    # Warning: genotype alleles must be sorted in ascending order!
+    if genotype[0] < 0:
+        return -1
+    index = 0
+    for i in range(len(genotype)):
+        a = genotype[i]
+        index += _comb_with_replacement(a, i + 1)
+    return index
+
+
+@numba_guvectorize(  # type: ignore
+    [
+        "void(int8[:,:], uint64[:], uint64[:])",
+        "void(int16[:,:], uint64[:], uint64[:])",
+        "void(int32[:,:], uint64[:], uint64[:])",
+        "void(int64[:,:], uint64[:], uint64[:])",
+    ],
+    "(n, k),(g)->(g)",
+)
+def _count_sorted_genotypes(
+    genotypes: ArrayLike, _: ArrayLike, out: ArrayLike
+) -> ArrayLike:  # pragma: no cover
+    # Warning: genotype alleles must be sorted in ascending order!
+    out[:] = 0
+    for i in range(len(genotypes)):
+        index = _sorted_genotype_index(genotypes[i])
+        if index >= 0:
+            out[index] += 1
+
+
+def count_variant_genotypes(
+    ds: Dataset,
+    *,
+    call_genotype: Hashable = variables.call_genotype,
+    merge: bool = True,
+) -> Dataset:
+    """Count the number of calls of each possible genotype, at each variant.
+
+    The "possible genotypes" at a given variant locus include all possible
+    combinations of the alleles at that locus, of size ploidy (i.e., all
+    multisets of those alleles with cardinality <ploidy>).
+
+    Parameters
+    ----------
+    ds
+        Dataset containing genotype calls.
+    call_genotype
+        Input variable name holding call_genotype as defined by
+        :data:`sgkit.variables.call_genotype_spec`.
+        Must be present in ``ds``.
+    merge
+        If True (the default), merge the input dataset and the computed
+        output variables into a single dataset, otherwise return only
+        the computed output variables.
+        See :ref:`dataset_merge` for more details.
+
+    Returns
+    -------
+    A dataset containing :data:`sgkit.variables.variant_genotype_count_spec`
+    of genotype counts with shape (variants, genotypes).
+
+    Warnings
+    --------
+    This method is not suitable for mixed-ploidy genotype calls. If it is
```

Comment:
Has there been discussion of whether this should be a hard error? I can imagine cases where a user hasn't realised this.

Reply:
Good call

> Does it make sense to raise an error from the numba functions if you hit a call that is < -1?

In the `call_genotype` spec we state "... or -2 to indicate a non allele in mixed ploidy datasets" which is a little ambiguous with respect to the non-mixed case. But probably best raise on < -1 for safety.